ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/recovery/PeriodicRecovery.java

```diff
@@ -186,31 +188,80 @@ public void shutdown (boolean async)
+     *     In case <code>isWaitForFinalRecovery() == true</code>, it is <b>very important</b> that, before invoking
+     *     <code>suspendScan</code>:
+     *     * The Transaction System (i.e. TxControl) gets disabled (i.e. no new transactions will be created)
+     *     * The Transaction Reaper finishes monitoring all in-flight transactions
+     * </p>
+     * @param async false if the calling thread should wait for any in-progress scan to complete before returning.
+     *              In case <code>isWaitForFinalRecovery() == true</code>, this parameter is override.
+     * @return the previous mode before attempting the suspension
      */
 
    public Mode suspendScan (boolean async)
    {
        synchronized (_stateLock)
        {
-           // only switch and kick everyone if we are currently ENABLED
            Mode currentMode = getMode();
 
+           /*
+            * isWaitForFinalRecovery checks whether suspension should delay until RecoveryModules
+            * implementing SuspendBlockingRecoveryModule recover all their transactions.
+            */
+           if (currentMode == Mode.ENABLED && recoveryPropertyManager.getRecoveryEnvironmentBean().isWaitForFinalRecovery()) {
+
+               /*
+                * At least, one clear scan (after the Transaction Manager has been shut down)
+                * should be completed to make sure that there are no transactions to recover.
+                */
+               doScanningWait();
+               doWork();
+
+               /*
+                * Now, it is finally possible to start checking if there are transactions that
+                * are still in need of recovery (or completion, in case of heuristics)
+                */
+               while (_blockSuspension) {
+                   /*
+                    * There might be another thread (the embedded thread or a user-request scan)
+                    * running the recovery cycle.
+                    * It would be possible to use again doScanningWait() here to wait the end of
+                    * the ongoing scan, but the best bet is to wait for the end of the next
+                    * recycle scan with doPeriodicWait()
+                    */
+                   doPeriodicWait();
+                   
+                   // I don't know the best place to make this comment but checking for subordinates could be a new recovery module wrapping something like https://github.com/jbosstm/narayana/blob/86182416bea64368ecdfc7e78767f798b15c14db/ArjunaJTS/jtax/classes/com/arjuna/ats/internal/jta/transaction/jts/jca/XATerminatorImple.java#L438C4-L438C4
+                   // but please know that these checks are called by external processes at any time and so that would need to be blocked while this suspension process continues and appropriately handled
+               }
+           }
+
+           // only switch and kick everyone if we are currently ENABLED
```

Comment:
I am not sure if this comment is relevant to the other changes you are making but adding it makes me wonder if it is? If it is relevant I think you would need to explain a bit more in the comment what the purpose is in relation to the changes that are being made, please?

Reply:
Ah-ha thank you that makes sense!

Reply:
Thanks. Solving this